controllers/lookup/k8s_client.go

```diff
@@ -0,0 +1,160 @@
+package lookup
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/datainfrahq/druid-operator/apis/druid/v1alpha1"
+	internalhttp "github.com/datainfrahq/druid-operator/pkg/http"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type K8sClient struct {
+	client client.Client
+}
+
+func NewK8sClient(client client.Client) K8sClient {
+	return K8sClient{
+		client: client,
+	}
+}
+
+func (c *K8sClient) FindLookups(ctx context.Context, reports map[types.NamespacedName]Report) (map[ClusterKey]map[LookupKey]Spec, error) {
+	specs := &v1alpha1.DruidLookupList{}
+	if err := c.client.List(ctx, specs); err != nil {
```

Comment:
It's preferable to make the list requests a more specific and then w can avoid a lot of logic.
We can list all lookups to a relevant Druid cluster.

Let's say you change the `Spec.DruidClusterName` to `Spec.DruidClusterRef` of type `ObjectReference` (not even `LocalObjectReference` if we want to be able to create lookups and clusters in a different namespace) which is more generic and native, you can than do this:

```go
var lookup DruidLookup
// logic to get the current reconciled DruidLookup (basically you need to get this object as a parameter)
druidLookups := &v1alpha1.DruidLookupList{}
opts := []client.ListOption { 
		client.InNamespace(lookup.Spec.DruidRef.Namespace),
		client.MatchingFields{"spec.druidRef.name": lookup.Spec.DruidRef.Name},
}
if err := c.client.List(ctx, specs, opts...); if err != nil {
    //do something
}
```

Reply:
As explained in the description of the PR, I landed on reconciliation strategy where every it is found out what lookups are desired and what lookups actually exists in the cluster. These two sets will form the following Venn-diagram.

![Venn-diagram of desired and actual lookups](https://github.com/datainfrahq/druid-operator/assets/4522613/a1ab3117-98a1-4929-a534-6556688d64e2)

In my experience of creating this PR this approach yielded less and more understandable code compared to do the bookkeeping required to reconcile one lookup at a time.

As I noted in the description, considering the entire context of every desired and actual lookup, requires the operator to fetch and compute on more data. I judged that the data would be small enough that it would be worth paying the extra cost in resource consumption to have what I believe to be a more robust and easily maintained code base.

But I must concede that this is my first foray into developing an operator. You are also the one who better knows this code base and what is idiomatic for it of us two. Thusly I defer the ultimate judgement of what the best strategy is to you.

Reply:
I understand and agree that reconciling each lookup in a separate loop is problematic. My suggestion here is to assume that more than one Druid object can exist on a Kubernetes cluster.  So we can limit the scope of the reconciliation to all lookups in a specific Druid cluster instead of all lookups of all Druid clusters.

Reply:
Also, could you add an explanation of it to the code? I think it's important that everyone that needs to edit/change this code in the future will have this in front of their eyes

Reply:
I started to look into only reconciling for one Druid cluster at a time. I however quickly realised that to know what cluster to reconcile in, I would have to look at the lookup resource the reconciliation request is for, as it is on that resource the druid cluster is specified.

To be able to reliably do that lookup, I realised that I'd would have to do the bookkeeping of finalisers (in-case the resource is deleted) and keeping track of what cluster the resource pointed at the last time the reconciler processed it (in-case what cluster the resource is pointing at is changed).

With that realisation, I'm concluding that even in a mixed approach there's a lot of bookkeeping, and I'm leaning towards giving the full bookkeeping approach another shot.

Any thoughts? 😄

Reply:
I went ahead and tried the bookkeeping approach again. I think I managed to produced something acceptable. I'll mark this as resolved for now. Feel free to re-open if you feel differently!